### ISSUE TITLE
Allow defining relationships in config schemas

### ISSUE THREAD

**Issue by radeksimko** [hashicorp/terraform#1255]
Given following the piece of code in `helper/schema/schema.go`

```go
func (m schemaMap) Input(
	input terraform.UIInput,
	c *terraform.ResourceConfig) (*terraform.ResourceConfig, error) {
	keys := make([]string, 0, len(m))
	for k, _ := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	for _, k := range keys {
...
```

all config options are sorted alphabetically and treated independently in the prompt. 
I can imagine cases where I would like the user to enter some config options first and eventually not ask for some other options based on the entered data.

Example?
Imagine following AWS provider config:

 - `credentials_provider` (`detect | iam | env | static`)
 - `credentials_file_path`
 - `credentials_file_profile_name`
 - `access_key`
 - `secret_key`
 - `security_token`

It is obvious that:

1. It would be annoying to ask user for all of these
2. Some options with certain values are making other ones superfluous (e.g. `credentials_provider` = `detect`, `iam` or `env` means that I should ignore all other options)

Would there be interest in such feature? If not, how would you suggest solving the problem with dependent config options above?

Related: #1049

**Comment by ketzacoatl** [comment]
In SaltStack, state files (formula, sort of like .tf resource definitions) are rendered as a template before being applied. In reading the source as a template, we can use all the power provided by the package rendering the template. For example, by default SaltStack uses the jinja renderer, so templates have all the power of jinja in them. In this way, we can _easily_ create and manipulate data structures internal to the template and otherwise used in rendering the template. Using Terraform's _count_ makes me want conditionals, iteration, and more powerful capabilities such as those in jinja, in Terraform's resource definitions. FWIW.

**Comment by radeksimko** [comment]
@ketzacoatl How does Jinja solution apply to this problem of dependencies between config options? Any example you can provide so we can be inspired?

**Comment by phinze** [comment]
@radeksimko reviewing some old issues here. We still don't have a facility for this, but I feel like the "user enters input interactively" use case is more or less a "first Terraform experience" sort of feature. Once you're using Terraform in anger, you're almost definitely configuring provider in some other way.

I wonder if it'd be easier to just make the most common provider configuration the interactive ones, rather than try to represent the tree of possible options in a sane way interactively. What do you think?

**Comment by radeksimko** [comment]
> use case is more or less a "first Terraform experience" sort of feature. Once you're using Terraform in anger, you're almost definitely configuring provider in some other way.

I wouldn't be so sure. It may be the case _now_ purely because of the way we deal with entered data (just throw it away after `apply`/`destroy` basically). If we'd be saving this anywhere, I think the story would differ (i.e. people would be using `tfvars` and `default`s less than they do now).

Related: https://github.com/hashicorp/terraform/issues/2550

> I wonder if it'd be easier to just make the most common provider configuration the interactive ones, rather than try to represent the tree of possible options in a sane way interactively.

So you mean basically having a very specific codified tree for provider configuration, but not try to solve the problem in runtime? Sounds like an interesting idea and probably solves the example I gave above. :+1:

**Comment by phinze** [comment]
>  If we'd be saving this anywhere, I think the story would differ (i.e. people would be using tfvars and defaults less than they do now).

Good point! I hadn't thought about this. :+1: 

So yeah it's easy to picture how a prompt opt-in or opt-out config boolean would work, but I can't picture how a "subset A or subset B" prompt config would work.

Today's behavior is that we prompt for the superset yeah? It'd be tricky to take away that behavior without surprising users who might be pressing `<enter>` a lot to get to the combination of attributes they use.

Hmm... more `[thinking]` necessary. :grinning:

**Comment by radeksimko** [comment]
Agreed it's not trivial, otherwise I would have came up with a PR instead of an issue already :smiley: 
I don't expect this to be solved in two lines of code.

Maybe we should think of inputs similar way we think about resources? i.e. in graphs? I'm not sure whether it helps or whether it makes the problem more complicated... I guess both.

either way... https://twitter.com/mitchellh/status/648506349807013888

**Comment by phinze** [comment]
> Maybe we should think of inputs similar way we think about resources? i.e. in graphs?

It's become a running HashiCorp joke that given any problem, eventually a graph will find its way somewhere into the solution. :grinning: 

Your points are well taken though - will continue to think about this.

**Comment by apparentlymart** [comment]
Just to throw in my two cents: in my world we've reduced user input to a bare minimum by deriving almost everything from data obtained from data-fetching resources like `terraform_remote_state` and `consul_keys`.

I'm personally more interested in fixing the warts surrounding that workflow and de-emphasizing direct configuration, so I agree with what @phinze said earlier about the user inputs being more of a "welcome to Terraform" flow than something I'd want to use in a "real" scenario, because I find it much more powerful to compute/derive these values than to hand-wire them, even if my hand wiring would end up stored in a file somewhere.

The only case where we ever use per-run user-supplied variables is to allow a single application config to be deployed separately to multiple environments, like I was describing in my comment over in #1964 a while back. We just have some root configs that have the provider configurations hard-coded (we use environment variables for credentials, so this just includes location-ish things like AWS region) and then all other configurations descend in some way from those root configs.

By all of this I'm not meaning to argue that we shouldn't improve the UX around user inputs like @radeksimko originally suggested... just hoping to throw another perspective into the mix in case it helps to think about the overall problem.

**Comment by scalp42** [comment]
The issue is that you don't want to rely on Consul to have Terraform working though (chicken/egg issue).

**Comment by apparentlymart** [comment]
Indeed... we use `terraform_remote_state` exclusively for the first bunch of layers until we get to the layer that creates the consul cluster. After that we use a mixture of `consul_keys` and `terraform_remote_state`, depending on whether the configuration also needs to be visible to the nodes at runtime... the nodes can of course only see what's persisted in Consul, not what was in the Terraform config.
